[Nested Tensor]Support SDPA math fallback for jagged layout nested tensor

YuqingJ opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116445


Support this fallback by converting the jagged layout NT to strided layout NT, and the convert the result back to jagged layout NT. 
This fallback might not be efficient since it uses unbind, contiguous and split.

YuqingJ committed:
[Nested Tensor]Support SDPA math fallback for jagged layout nested tensor

[ghstack-poisoned]

YuqingJ committed:
Update on "[Nested Tensor]Support SDPA math fallback for jagged layout nested tensor"



Support this fallback by converting the jagged layout NT to strided layout NT, and the convert the result back to jagged layout NT. 
This fallback might not be efficient since it uses unbind, contiguous and split.



[ghstack-poisoned]

YuqingJ committed:
Update on "[Nested Tensor]Support SDPA math fallback for jagged layout nested tensor"



Support this fallback by converting the jagged layout NT to strided layout NT, and the convert the result back to jagged layout NT. 
This fallback might not be efficient since it uses unbind, contiguous and split.



[ghstack-poisoned]

YuqingJ commented on test/test_nestedtensor.py:
```diff
@@ -3622,14 +3614,15 @@ def check_forward_backward():
-                attn_nt = torch.nn.functional.scaled_dot_product_attention(q_nt_t, k_nt_t, v_nt_t).transpose(1, 2)
+            # Math fallback doesn't work with bfloat16 on CUDA because
+            # "group_gemm_dispatch" not implemented for 'BFloat16'
+            if not (device == "cuda" and dtype == torch.bfloat16):
```
Any suggestion on how to bypass this correctly?

soulitzer replied:
Skipping here should be fine

YuqingJ replied:
but the test still fails here (https://github.com/pytorch/pytorch/actions/runs/7335911393/job/19975043647). maybe the device doesn't call "cuda" exactly sometimes?

soulitzer replied:
Hmm that is strange, "cuda" should be correct. Maybe try printing out the device?

YuqingJ replied:
it turns out to be "cuda:0". now it works.

soulitzer commented on torch/nested/_internal/sdpa.py:
```diff
@@ -736,9 +736,30 @@ def jagged_scaled_dot_product_attention(
+        )
+        attn_out = attn_out.transpose(1, 2)
+
+        return attn_out
```
probably better to put this in NestedTensor's `__torch_function__` instead

adding this should be fine to land to unblock, but we'll eventually want to dedupe with https://github.com/pytorch/pytorch/pull/115749

YuqingJ replied:
Do you mean add the "_scaled_dot_product_attention_math" in jagged_torch_function in ops.py instead? Wondering why it is better to put it there? It seems to add an extra routing?

soulitzer replied:
We actually already have a route for sdpa there and the math path, see:  https://github.com/pytorch/pytorch/blob/main/torch/nested/_internal/sdpa.py#L738

Do you know what happens today if we reach there?

YuqingJ replied:
This is where I am adding the logic, no? Did I miss something?
Today, this path will cause "RuntimeError: scaled_dot_product_attention: If inputs are nested tensors they must be contiguous".

soulitzer replied:
No, I just misread lol

YuqingJ committed:
Update on "[Nested Tensor]Support SDPA math fallback for jagged layout nested tensor"



Support this fallback by converting the jagged layout NT to strided layout NT, and the convert the result back to jagged layout NT. 
This fallback might not be efficient since it uses unbind, contiguous and split.



[ghstack-poisoned]

YuqingJ committed:
Update on "[Nested Tensor]Support SDPA math fallback for jagged layout nested tensor"



Support this fallback by converting the jagged layout NT to strided layout NT, and the convert the result back to jagged layout NT. 
This fallback might not be efficient since it uses unbind, contiguous and split.



[ghstack-poisoned]

soulitzer commented on torch/nested/_internal/sdpa.py:
```diff
@@ -736,9 +736,30 @@ def jagged_scaled_dot_product_attention(
+        attn_out = attn_out.transpose(1, 2)
+        attn_out = torch.nested.as_nested_tensor(
+            list(attn_out.unbind()), layout=torch.jagged
+        )
```
Could we reshape the output buffer to 2D and then use ViewNestedFromBuffer?
If we construct with as_nested_tensor, it would acquire a new singleton_int and you wouldn't be able to do elementwise binary ops for example because we would consider the two tensors to have different jaggedness

YuqingJ replied:
I think that's a bit tricky to do so. The c++ NT values doesn't change after transpose. So I can't transpose it and then reshape it to be the buffer for ViewNestedFromBuffer.  

(More context: the output of _scaled_dot_product_attention_math is a 4D C++ Nested tensor with shape like [B, D1, j0, D2].  The goal is to convert this C++ NT to PY NT. Since ViewNestedFromBuffer assume the jagged dimension is the 2nd dimension, I need to transpose the [B, D1, j0, D2] to  [B, j0, D1, D2] first. But since transpose doesn't change the values() in C++ NT, I can't get the transposed values.)

Any other ways to work around it? By the way, for my use case, we patch the attention output before doing any elementwise binary operations, so this is not a blocker for us.

soulitzer replied:
What if you got the values from cpp, makes it 2D, and transpose the values yourself and then call contiguous before calling ViewNestedFromBuffer.

soulitzer commented on torch/nested/_internal/sdpa.py:
```diff
@@ -736,9 +736,30 @@ def jagged_scaled_dot_product_attention(
         )[0]
+        attn_out = attn_out.transpose(1, 2)
+        attn_out = torch.nested.as_nested_tensor(
+            list(attn_out.unbind()), layout=torch.jagged
```
would be cool to be able to pass in offsets to this in the future, to ensure the shapes compare equal later on, but this is fine for now

YuqingJ replied:
yeah, or maybe have a better API to convert C++ NT to PY NT if there are more use cases like this.

YuqingJ commented:
@pytorchbot merge

YuqingJ committed:
Update on "[Nested Tensor]Support SDPA math fallback for jagged layout nested tensor"



Support this fallback by converting the jagged layout NT to strided layout NT, and the convert the result back to jagged layout NT. 
This fallback might not be efficient since it uses unbind, contiguous and split.



[ghstack-poisoned]

YuqingJ committed:
Update on "[Nested Tensor]Support SDPA math fallback for jagged layout nested tensor"



Support this fallback by converting the jagged layout NT to strided layout NT, and the convert the result back to jagged layout NT. 
This fallback might not be efficient since it uses unbind, contiguous and split.



[ghstack-poisoned]

YuqingJ committed:
Update on "[Nested Tensor]Support SDPA math fallback for jagged layout nested tensor"



Support this fallback by converting the jagged layout NT to strided layout NT, and the convert the result back to jagged layout NT. 
This fallback might not be efficient since it uses unbind, contiguous and split.



[ghstack-poisoned]

YuqingJ committed:
Update on "[Nested Tensor]Support SDPA math fallback for jagged layout nested tensor"



Support this fallback by converting the jagged layout NT to strided layout NT, and the convert the result back to jagged layout NT. 
This fallback might not be efficient since it uses unbind, contiguous and split.



[ghstack-poisoned]

soulitzer reviewed (approved):
LGTM

YuqingJ commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but i see it fails ROCm test in trunk due to an unsupported use case https://hud.pytorch.org/pytorch/pytorch/commit/e55a778cbb518e54c5afa5b8107b352746d7f41a' -c nosignal

I think it's easier to revert and reland this.  I have added `ciflow/rocm` into your PR to get the job running.

YuqingJ committed:
Update on "[Nested Tensor]Support SDPA math fallback for jagged layout nested tensor"



Support this fallback by converting the jagged layout NT to strided layout NT, and the convert the result back to jagged layout NT. 
This fallback might not be efficient since it uses unbind, contiguous and split.



[ghstack-poisoned]

YuqingJ commented:
@pytorchbot merge
